lib/lint.ml

```diff
@@ -8,459 +8,33 @@ let () = Lwt_preemptive.init 0 1 (fun _errlog -> ()) (* NOTE: Lwt_preemptive is
-              (pkg, PrefixConflictClassMismatch
-                (WrongPrefix { conflict_class; required_prefix=prefix })) :: errors
-        )
-        errors
-        conflict_classes
-
-    let check_prefix_conflict_class_mismatch ~errors ~pkg opam =
-      let conflict_classes = List.map OpamPackage.Name.to_string @@ OpamFile.OPAM.conflict_class opam in
-      let name = OpamPackage.name_to_string pkg in
-      let errors = check_prefix_without_conflict_class ~errors ~pkg name conflict_classes in
-      check_conflict_class_without_prefix ~errors ~pkg name conflict_classes
-  end
-
-  let opam_lint ~errors ~pkg opam =
-    OpamFileTools.lint ~check_upstream:true opam |>
-    List.fold_left (fun errors x -> (pkg, OpamLint x) :: errors) errors
-
-  let of_dir ~host_os ~master ~job ~packages cwd =
+  let of_dir ~master ~job ~packages cwd =
     let master = Current_git.Commit.hash master in
     exec ~cwd ~job [|"git"; "merge"; "-q"; "--"; master|] >>/= fun () ->
-    get_packages ~cwd >>= fun existing_packages ->
-    Lwt_list.fold_left_s (fun errors (pkg, kind) ->
-      match kind with
-      | Analyse.Analysis.Deleted ->
-          Lwt.return errors (* TODO *)
-      | Analyse.Analysis.(New _ | Unavailable | SignificantlyChanged | InsignificantlyChanged) ->
-          get_opam ~cwd pkg >>= fun opam ->
-          let errors = check_name_field ~errors ~pkg opam in
-          let errors = check_version_field ~errors ~pkg opam in
-          let errors = check_dune_subst ~errors ~pkg opam in
-          let errors = check_checksums ~errors ~pkg opam in
-          let errors = check_no_pin_depends ~errors ~pkg opam in
-          let errors = check_no_extra_files ~errors ~pkg opam in
-          let errors = Prefix.check_prefix_conflict_class_mismatch ~errors ~pkg opam in
-          check_dune_constraints ~host_os ~errors ~pkg opam >>= fun errors ->
-          (* Check directory structure correctness *)
-          scan_dir ~cwd errors pkg >>= fun errors ->
-          let errors = opam_lint ~errors ~pkg opam in
-          match kind with
-          | New Package ->
-            let errors = check_name_collisions ~errors ~pkg existing_packages in
-            Lwt.return (Prefix.check_name_restricted_prefix ~errors ~pkg)
-          | _ -> Lwt.return errors
-    ) [] packages
+    let changed = List.filter_map (fun (pkg, change) ->
+        match change with
+        | Analyse.Analysis.(New Release | Unavailable | SignificantlyChanged | InsignificantlyChanged ) -> Some (OpamPackage.to_string pkg)
+        | _ -> None
+      ) packages in
+    let new_ = List.filter_map (fun (pkg, change) ->
+        match change with
+        | Analyse.Analysis.(New Package) -> Some (OpamPackage.to_string pkg)
+        | _ -> None
+      ) packages in
+    let cmd = ["opam-ci-check"; "lint"; "-r"; "."] @
+              (if changed <> [] then ["-c"; String.concat "," changed ] else []) @
+              (if new_ <> [] then ["-n"; String.concat "," new_ ] else [])
```

Comment:
I suggest moving the parsing of the `changed`/`new_` values into CLI flags be into the extraction step. IMO, it makes for a cleaner separation of concerns and makes the CLI invocation clearer.

This could be done in several ways, but would be something like:

```suggestion
    let changed =
      packages
      |> List.filter_map (fun (pkg, change) ->
          match change with
          | Analyse.Analysis.(New Release | Unavailable | SignificantlyChanged | InsignificantlyChanged ) -> Some (OpamPackage.to_string pkg)
          | _ -> None) 
      |> function
      | [] -> []
      | changed_packages -> ["--changed-packages"; String.concat "," changed_packages]
    in
    let new_ =
      packages
      |> List.filter_map (fun (pkg, change) ->
          match change with
          | Analyse.Analysis.(New Package) -> Some (OpamPackage.to_string pkg)
          | _ -> None) 
      |> function
      | [] -> []
      | new_packages -> ["--newly-published"; String.concat "," new_packages]
    in
    let cmd = ["opam-ci-check"; "lint"; "-r"; "."] @ changed @ new_
```

This way we only have to reason about what gets computed for these flags in one place.

Reply:
Makes sense. Changed!